src/vos/vos_pool.c

```diff
@@ -617,7 +632,7 @@ vos_pmemobj_create(const char *path, uuid_t pool_id, const char *layout,
 
 static int
 vos_pmemobj_open(const char *path, uuid_t pool_id, const char *layout, unsigned int flags,
-		 struct umem_pool **ph)
+		 void *metrics, struct umem_pool **ph)
```

Comment:
Instead of changing this vos_pmemobj_open(), why don't we just change callback of 'so_wal_replay' to accept an optional 'struct umem_replay_stats *stats'?

Reply:
If the pool metrics reference is passed as argument to bio_wal_replay, that would resolve all your concerns mentioned above this comment, right? No 'struct bio_wal_rp_stats' in mc; no bio_wal_set_metrics() BIO API and so on - none would be needed!

Reply:
Also no problem with removing per-engine WAL replay stats. Though those could be useful in some specific user cases when the after-mortals are queried, the per-engine stats are mostly redundant.

Reply:
As you mentioned that BIO is on top of UMEM, lower layer exporting API or data structure to higher layer looks not a layer violation to me, in contrast, lower layer should know nothing about higher layer.

You shouldn't pass VOS data structure (or VOS metrics reference) to BIO, since VOS is on top of BIO.